src/server.ts

```diff
@@ -101,16 +123,372 @@ If the user asks to schedule a task, use the schedule tool to schedule the task.
+        }
+      );
+
+      const authCheck = (await authCheckResponse.json()) as {
+        authenticated: boolean;
+      };
+
+      return c.json({
+        success: true,
+        authenticated: authCheck.authenticated,
+      });
+    }
+
+    // Regular authentication flow
+    if (!providedKey) {
+      return c.json({ error: "providedKey is required" }, 400);
+    }
+
+    const expectedKey = c.env.ADMIN_SECRET; // Use the environment variable from Cloudflare
+
+    // Send validation request to the Durable Object
+    const authResponse = await sessionTracker.fetch(
+      "https://dummy-host/validate-session-auth",
+      {
+        method: "POST",
+        headers: { "Content-Type": "application/json" },
+        body: JSON.stringify({ sessionId, providedKey, expectedKey }),
+      }
+    );
+
+    const authResult = (await authResponse.json()) as {
+      success: boolean;
+      authenticated: boolean;
+      authenticatedAt?: string;
+      error?: string;
+    };
+
+    if (authResult.success && authResult.authenticated) {
+      return c.json({
+        success: true,
+        authenticated: true,
+        authenticatedAt: authResult.authenticatedAt,
+      });
+    }
+    return c.json(
+      {
+        success: false,
+        authenticated: false,
+        error: "Invalid admin key",
+      },
+      401
+    );
+  } catch (error) {
+    console.error("Error authenticating session:", error);
+    return c.json({ error: "Internal server error" }, 500);
+  }
+});
+
+// PDF Upload URL Route (for presigned uploads)
+app.post("/pdf/upload-url", async (c) => {
+  try {
+    const { sessionId, fileName, fileSize } = await c.req.json();
+
+    if (!sessionId || !fileName) {
+      return c.json({ error: "sessionId and fileName are required" }, 400);
+    }
+
+    // Check if session is authenticated
+    const sessionIdObj = c.env.SessionFileTracker.idFromName(sessionId);
+    const sessionTracker = c.env.SessionFileTracker.get(sessionIdObj);
+
+    const authCheckResponse = await sessionTracker.fetch(
+      "https://dummy-host/is-session-authenticated",
+      {
+        method: "GET",
+      }
+    );
+
+    const authCheck = (await authCheckResponse.json()) as {
+      authenticated: boolean;
+    };
+    if (!authCheck.authenticated) {
+      return c.json({ error: "Session not authenticated" }, 401);
+    }
+
+    // Generate unique file key
+    const fileKey = `uploads/${sessionId}/${crypto.randomUUID()}-${fileName}`;
+
+    // Generate direct upload URL to R2 bucket
+    // This creates a URL that uploads directly to R2, bypassing the worker
+    const uploadUrl = `/pdf/upload/${fileKey}`;
+
+    // Add file metadata to SessionFileTracker
+    await sessionTracker.fetch("https://dummy-host/add-file", {
+      method: "POST",
+      headers: { "Content-Type": "application/json" },
+      body: JSON.stringify({
+        sessionId,
+        fileKey,
+        fileName,
+        fileSize: fileSize || 0,
```

Comment:
0 is probably not a great default, let's use one of the libraries i have in a comment above + ensure we know the exactly fliesize. i also recommend we gray out upload button on `fileSize: 0`